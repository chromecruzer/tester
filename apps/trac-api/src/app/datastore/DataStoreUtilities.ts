import fs from 'fs-extra';
import util from "util";

const dump = (obj, depth = null) => util.inspect(obj, {depth, colors: true});

export default class DataStoreUtilities {
  fs = fs;
  public async retrieve(filepath) {
    console.log(`reading file from client`)  // added by me 
    return this.fs.readJson(filepath)
      .catch(e => Promise.reject(new Error(`File ${filepath} could not be read. ${e.message}`)));
  }
  
  public async store(filepath, json) {
    // console.log(`writing content ${dump(json)} to filepath ${filepath}`)
    return this.fs.writeJson(filepath, json)
      .catch(e => Promise.reject(new Error(`Cannot save file ${filepath} because ${e.message}`)));
  }
  public async check(filepath) {
    const result = await this.fs.pathExists(filepath);
    // console.log(`file ${filepath} exists ${result}`);
    return result;
  }
  public async remove(filepath) {
    return this.fs.remove(filepath)
      .catch(e => Promise.reject(new Error(`Cannot remove ${filepath} because ${e.message}`)));
  }
  public async mkdir(dir) {
    return this.fs.ensureDir(dir)
      .catch(err => Promise.reject(new Error(`Cannot create upload folder "${dir}" in ${dir} because ${err.message}`)));
  }
}
