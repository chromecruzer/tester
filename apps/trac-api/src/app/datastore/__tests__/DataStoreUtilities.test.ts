import DataStoreUtilities from "../DataStoreUtilities";

describe('DataStoreUtilities', () => {
  let mockFs, dsu;
  beforeEach(() => {
    mockFs = {
      ensureDir: jest.fn().mockReturnValue(Promise.resolve()),
      ensureFile: jest.fn().mockReturnValue(Promise.resolve(true)),
      remove: jest.fn().mockReturnValue(Promise.resolve()),
      writeJson: jest.fn().mockReturnValue(Promise.resolve()),
      readJson: jest.fn().mockReturnValue(Promise.resolve({}))
    };
    dsu = new DataStoreUtilities();
    dsu.fs = mockFs;
  });
  describe('retrieve', () => {
    it('should call readJson with the file path', () => {
      const expected = 'thisIsAFilePath'
      dsu.retrieve(expected);
      expect(mockFs.readJson).toBeCalledWith(expected);
    });
    it('should reject with an error when readJson fails', () => {
      const testCall = () => dsu.retrieve('whoCares');
      mockFs.readJson.mockImplementation(() => Promise.reject(new Error('This did not work')));
      expect(testCall).rejects.toMatchSnapshot();
    });
  });
  describe('store', () => {
    it('should call writeJson with the file path and content', () => {
      const expected = ['thisIsAFilePath', {content: 'this is content'}];
      dsu.store(...expected);
      expect(mockFs.writeJson).toBeCalledWith(...expected, {"flag": "w", "spaces": 4});
    });
    it('should reject with an error when writeJson fails', () => {
      const testCall = () => dsu.store('whoCares', {none:'no content'});
      mockFs.writeJson.mockImplementation(() => Promise.reject(new Error('This did not work')));
      expect(testCall).rejects.toMatchSnapshot();
    });
  });
  describe('check', () => {
    it('should return false if file is absent', () => {
      const expected = 'thisIsAFilePath'
      mockFs.ensureFile.mockReturnValue(Promise.reject(false));
      return dsu.check(expected).then(actual => {
        expect(mockFs.ensureFile).toBeCalledWith(expected);
        expect(actual).toBeFalsy();
      });
    });
    it('should return true if the file is present', () => {
      const expected = 'thisIsAFilePath'
      return dsu.check(expected).then(actual => {
        expect(mockFs.ensureFile).toBeCalledWith(expected);
        expect(actual).toBeTruthy();
      });
    });
  });
  describe('remove', () => {
    it('should call remove with the file path', () => {
      const expected = 'thisIsAFilePath'
      dsu.remove(expected);
      expect(mockFs.remove).toBeCalledWith(expected);
    });
    it('should reject with an error when remove fails', () => {
      const testCall = () => dsu.remove('whoCares');
      mockFs.remove.mockImplementation(() => Promise.reject(new Error('This did not work')));
      expect(testCall).rejects.toMatchSnapshot();
    });
  });
  describe('mkdir', () => {
    it('should call ensureDir with the file path', () => {
      const expected = 'thisIsADirPath'
      dsu.mkdir(expected);
      expect(mockFs.ensureDir).toBeCalledWith(expected);
    });
    it('should reject with an error when ensureDir fails', () => {
      const testCall = () => dsu.mkdir('whoCares');
      mockFs.ensureDir.mockImplementation(() => Promise.reject(new Error('This did not work')));
      expect(testCall).rejects.toMatchSnapshot();
    });
  });
});
