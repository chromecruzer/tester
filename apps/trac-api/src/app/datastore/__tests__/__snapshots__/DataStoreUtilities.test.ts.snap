// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DataStoreUtilities mkdir should reject with an error when ensureDir fails 1`] = `[Error: Cannot create upload folder "whoCares" in whoCares because This did not work]`;

exports[`DataStoreUtilities remove should reject with an error when remove fails 1`] = `[Error: Cannot remove whoCares because This did not work]`;

exports[`DataStoreUtilities retrieve should reject with an error when readJson fails 1`] = `[Error: File whoCares could not be read. This did not work]`;

exports[`DataStoreUtilities store should reject with an error when writeJson fails 1`] = `[Error: Cannot save file whoCares because This did not work]`;
